components/script/indexed_db.rs

```diff
@@ -186,158 +193,204 @@ pub fn convert_value_to_key_range(
+                }
 
-    rooted!(in(*cx) let mut target_object = ptr::null_mut::<JSObject>());
-    rooted!(in(*cx) let mut current_val = *value);
-    rooted!(in(*cx) let mut object = ptr::null_mut::<JSObject>());
+                // Step 1.3.7. Increase i by 1.
+                // Done by for loop with enumerate()
+            }
 
-    let mut target_object_prop_name: Option<String> = None;
+            // Step 1.4. Return result.
+            result.safe_to_jsval(cx, return_val);
+        },
+        KeyPath::String(key_path) => {
+            // Step 2. If keyPath is the empty string, return value and skip the remaining steps.
+            if key_path.is_empty() {
+                return_val.set(*value);
+                return Ok(());
+            }
 
-    match key_path {
-        KeyPath::String(path) => {
-            // Step 3
-            let path_as_string = path.to_string();
-            let mut tokenizer = path_as_string.split('.').peekable();
-
-            while let Some(token) = tokenizer.next() {
-                if target_object.get().is_null() {
-                    if token == "length" && tokenizer.peek().is_none() && current_val.is_string() {
-                        rooted!(in(*cx) let input_val = current_val.to_string());
-                        unsafe {
-                            let string_len = JS_GetStringLength(*input_val) as u64;
-                            string_len.safe_to_jsval(cx, return_val);
-                        }
-                        break;
+            // NOTE: Use current_value, instead of value described in spec, in the following steps.
+            rooted!(in(*cx) let mut current_value = *value);
+
+            // Step 3. Let identifiers be the result of strictly splitting keyPath on U+002E
+            // FULL STOP characters (.).
+            // Step 4. For each identifier of identifiers, jump to the appropriate step below:
+            for identifier in key_path.split('.') {
+                // If Type(value) is String, and identifier is "length"
+                if identifier == "length" && current_value.is_string() {
+                    // Let value be a Number equal to the number of elements in value.
+                    rooted!(in(*cx) let string_value = current_value.to_string());
+                    unsafe {
+                        let string_length = JS_GetStringLength(*string_value) as u64;
+                        string_length.safe_to_jsval(cx, current_value.handle_mut());
                     }
-
-                    if !current_val.is_object() {
-                        // FIXME:(rasviitanen) Return a proper error
-                        return;
+                }
+                // If value is an Array and identifier is "length"
+                else if identifier == "length" {
+                    unsafe {
+                        let mut is_array = false;
+                        if IsArrayObject(*cx, current_value.handle(), &mut is_array) {
```

Comment:
this returns Error::Data when IsArrayObject succeeds. It should return an error only when the API call fails, right?